spanconfig: replace roachpb.SpanConfigEntry in package spanconfig

arulajmani opened the pull request:
See individual commits for details. 

Polished up and carved out from https://github.com/cockroachdb/cockroach/pull/76138. 

In the next patch, we'll modify the existing span configuration RPC arguments to something similar in the draft PR above. Then, we'll make `spanconfig.Target` a union over system targets and spans (instead of the alias for roachpb.Spans as in this PR). We'll also add all the encoding and decoding methods for system targets that we've seen in prior draft patches.

arulajmani committed:
Revert "roachpb,server,spanconfig: introduce RPCs for SystemSpanConfigs"

Revert the RPCs we added for system span configurations; we'll instead
modify the existing RPCs to deal with system span configurations in a
future patch.

This reverts commit 5177417a0bffe68ee1ca1ee310ec4717613ae0b7.

Release note (<category, see below>): <what> <show> <why>

arulajmani committed:
spanconfig: replace roachpb.SpanConfigEntry in package spanconfig

This patch removes the usage of roachpb.SpanConfigEntry from the
spanconfig package. Going forward, we'll only use
roachpb.SpanConfigEntry in RPCs. We want to decouple types that we
use inside the package with RPCs in preperation for system span
configurations.

We instead introduce a new spanconfig.Record type to tie together
a spanconfig.Target and a spanconfig.Record. For now, we only allow
targeting spans. Once we introduce system span configurations, we'll
make be able to make room for system targets as well. System targets
will allow tenants to target their entire keyspace and the host tenant
to target particular secondary tenants/the entire cluster.

Crucially, spanconfig.Target contains encoding and decoding methods
which we make use of when writing to and reading from
system.span_configurations.

Ripping out roachpb.SpanConfigEntry ended up touching most components
in the package.

Release note: None

arulajmani committed:
roachpb: remove system span config protos

We'll instead use roachpb.SpanConfig to represent system span configs
as well; in the future, we'll just wrap it in a go type an do some
validation.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76213)
<!-- Reviewable:end -->

arulajmani committed:
spanconfigstore: pull out the span config storage into internal struct

This commit pulls out what used to be the spanconfig.Store logic into
as separate spanconfigstore.spanConfigStore struct to exclusively track
span <-> config pairs. This also makes room for adding another internal
struct to track system span configs in a future patch.

Release note: None

irfansharif commented on pkg/spanconfig/spanconfigstore/spanconfigstore.go:
```diff
@@ -0,0 +1,447 @@
+}
+
+// copy returns a copy of the spanConfigStore.
+func (s *spanConfigStore) copy(ctx context.Context) *spanConfigStore {
```
I'm not going to hold up the merge over it, but I don't think this commit is a positive change. Thanks for separating it out into a stand-alone commit -- I think we should drop it. We have three variants now, and one in the future.

- spanconfig.Store;
- spanconfigstore.Store;
- spanconfigstore.spanConfigStore;
- (in the future) spanconfigstore.systemSpanConfigStore

This kind of spread, if anything, makes it difficult to talk about things meaninfully. Looking at the tests, I don't think the separation is also helping much. I understand the intent here was to make room for the stand-alone spanconfigstore.systemSpanConfigStore, but that too I don't think needs to be a wholly distinct type when instead it could be a set of fields on spanconfigstore.Store and appropriately named methods. The interface we care about for all the three types are spanconfig.StoreWriter (with it's diff-ing contract), primarily, and I think it's possible to test each component by interfacing through just the StoreWriter interface. Want to test spanconfigstore.systemSpanConfigStore? Just populate using the right kind of span.Targets. What to test spanconfigstore.systemSpanConfigStore? Do the same thing. Want to test the top-level spanconfigstore.Store? Again, just do the same thing.

Generally IMO when it's possible to test everything we care about through the high-level interface (which is the case here), I think we should. Peeking into the internals instead makes for test code that's going to be subject to more change over time than necessary.

irfansharif reviewed (approved):
These new types look great! I think framing the changes to this package in terms of spanconfig.{Target,Record} and in the future generalizing spanconfig.Target to also encompass the "special" targets is very clear IMO -- I dig it. I've left mostly minor comments below, with a recommendation that we abandon the third commit. I also have a question about the continued existence of roachpb.SpanConfigEntry, but nothing worth holding this PR on.

arulajmani reviewed (commented):
TFTR! Will bors on green. 

edit -- missed a few comments above because they were over-written in a subsequent commit. Will merge on green and once those are resolved.

arulajmani commented on pkg/ccl/migrationccl/migrationsccl/seed_tenant_span_configs_external_test.go:
```diff
@@ -124,7 +124,7 @@ func TestSeedTenantSpanConfigs(t *testing.T) {
 
 	{ // Ensure that no span config entries are to be found
-		entries, err := scKVAccessor.GetSpanConfigEntriesFor(ctx, []roachpb.Span{
+		entries, err := scKVAccessor.GetSpanConfigRecords(ctx, []roachpb.Span{
```
Done.

arulajmani commented on pkg/spanconfig/spanconfigstore/store.go:
```diff
@@ -178,60 +201,70 @@ func (s *Store) ForEachOverlapping(
 		}
 	}
 
+	if err := validateIntervalTreeApplyArgs(updates...); err != nil {
```
Yup, good call, done!

arulajmani commented on pkg/spanconfig/spanconfigstore/store.go:
```diff
@@ -145,27 +149,46 @@ func (s *Store) Apply(
-	_ = s.ForEachOverlapping(ctx, keys.EverythingSpan, func(entry roachpb.SpanConfigEntry) error {
-		clone.Apply(ctx, false /* dryrun */, spanconfig.Update(entry))
+	_ = s.forEachOverlappingRLocked(keys.EverythingSpan, func(entry spanConfigEntry) error {
+		_, _, err := clone.applyInternal(false /* dryrun */, spanconfig.Update{
```
Done!

arulajmani commented on pkg/spanconfig/spanconfigstore/store_test.go:
```diff
@@ -181,9 +240,9 @@ func TestRandomized(t *testing.T) {
 		switch op {
 		case "set":
-			return spanconfig.Addition(sp, conf)
+			return spanconfig.Addition(spanconfig.MakeSpanTarget(sp), conf)
```
Done in the intermediate commit. Though, this does speak to my point of this being a test for the internal struct and not the interface/struct that implements the interface.

arulajmani commented on pkg/spanconfig/spanconfig.go:
```diff
@@ -24,24 +24,25 @@ import (
+	// GetSpanConfigRecords returns the span configurations that overlap with
 	// the given spans.
-	GetSpanConfigEntriesFor(
+	GetSpanConfigRecords(
```
Changed.

arulajmani commented:
It's green! TFTR, and thanks for brainstorming a bunch of this stuff to get to this point. I echo your sentiments around the new types! 

bors r+
